examples/zonal_private_iap_bastion/network.tf

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module "gke-network" {
+  source       = "terraform-google-modules/network/google"
+  project_id   = var.project_id
+  network_name = "my-network"
+
+  subnets = [
+    {
+      subnet_name           = "my-subnet"
+      subnet_ip             = "10.0.0.0/24"
+      subnet_region         = var.region
+      subnet_private_access = "true"
+    },
+  ]
+
+  secondary_ranges = {
+    "my-subnet" = [
+      {
+        range_name    = "my-network-${var.cluster_name}-pod-range"
+        ip_cidr_range = "10.1.0.0/16"
+      },
+      {
+        range_name    = "my-network-${var.cluster_name}-service-range"
+        ip_cidr_range = "10.2.0.0/20"
+      },
+  ] }
+}
+
+module "cloud-nat" {
+  source        = "terraform-google-modules/cloud-nat/google"
+  project_id    = var.project_id
+  region        = var.region
+  create_router = "true"
+  router        = "${var.cluster_name}-cloud-nat"
+  network       = module.gke-network.network_name
+}
```

Comment:
I found I had to add an explicit route for destination 0.0.0.0/0 in order to take advantage of the NAT instance. I didn't completely tear down the host project, however, so maybe the original firewall configuration had left some configuration changes behind?  I did destroy all the subnets and recreate them, this time skipping the creation of any firewall rules, and the bastion still couldn't access the public internet.  Nor could it get to nodes in my gke subnet.  The routes existed, but it needed a firewall rule granting ingress permission.  Examples provided below.

```
resource "google_compute_route" "bastion-egress" {
  project          = var.project_id
  network          = module.gke-network.network_name
  name             = "bastion-egress"
  description      = "Allow bastion to talk to the internet"

  # latest bastion module allows setting of tags.  Not released, but merged.  
  # Otherwise use service_account?
  tags             = var.tags

  dest_range       = "0.0.0.0/0"
  next_hop_gateway = "default-internet-gateway"
}

resource "google_compute_firewall" "bastion-ingress" {
  # required for 'enable-logging'
  provider       = "google-beta"
  name           = "ingress-bastion"
  description    = "Allow inbound traffic from bastion host"
  network        = module.gke-network.network_name
  project        = var.project_id
  direction      = "INGRESS"
  enable_logging = true

  # use service account if you don't have tags
  source_tags    = ["bastion"]

  dynamic "allow" {
    for_each = ["tcp", "udp", "icmp"]
    content {
      protocol = allow.value
    }
  }
}
```

Reply:
Another data point.  When I added a bunch of services to the cluster, requiring vertical auto_scaling, the node pool size was bumped up to 4 (from 2), and the new instances launched, but until I explicitly added a route to the public internet for nodes in the cluster, the new nodes were not detected by the kubernetes cluster and pods remained unschedulable.  As soon as I added a route to 0.0.0.0./0 for gke nodes, the additional nodes were detected and pods were scheduled. 

I had such a route when launching the cluster, but eventually swapped it for one which only allowed the bastion to go out to 0.0.0.0/0.  Then I tried to scale the cluster up, which didn't work until I added the route back.  I suspect that this implies that private_access requires more configuration than simply setting subnet_private_access to true in the network module, though I don't know what else is required.  The docs sure seem to imply that it should just work unless you've overwritten the default route (https://cloud.google.com/vpc/docs/configure-private-google-access) so maybe this is another artifact of my prior use of the firewall module without destroying the host project and recreating it, but I did destroy all subnets and recreate them.

My guess would be that it wants me to configure DNS in the VPC with a cname from *.googleapis.com to restricted.googleapis.com in order to make this work without explicitly routing to 0.0.0.0/0 for any gke nodes (in order to go through NAT, but that implies it isn't using the private access without such a cname).

Reply:
It looks like I forgot to add the subnets from the original example. Will add and test.